pytest/tests/loadtest/locust/common/base.py

```diff
@@ -0,0 +1,324 @@
+
+
+class TransferNear(Transaction):
+
+    def __init__(self, sender, recipient_id, how_much=2.0):
+        super().__init__()
+        self.recipient_id = recipient_id
+        self.sender = sender
+        self.how_much = how_much
+
+    def sign_and_serialize(self, block_hash):
+        sender = self.sender
+        logger.debug(
+            f"sending {self.how_much} NEAR from {sender.key.account_id} to {self.recipient_id}"
+        )
+        return transaction.sign_payment_tx(sender.key, self.recipient_id,
+                                           int(self.how_much * 1E24),
+                                           sender.use_nonce(), block_hash)
+
+
+class CreateSubAccount(Transaction):
+
+    def __init__(self, sender, sub_key, balance=50.0):
+        super().__init__()
+        self.sender = sender
+        self.sub_key = sub_key
+        self.balance = balance
+
+    def sign_and_serialize(self, block_hash):
+        sender = self.sender
+        sub = self.sub_key
+        logger.debug(f"creating {sub.account_id}")
+        return transaction.sign_create_account_with_full_access_key_and_balance_tx(
+            sender.key, sub.account_id, sub, int(self.balance * 1E24),
+            sender.use_nonce(), block_hash)
+
+
+class NearUser(HttpUser):
+    abstract = True
+    id_counter = 0
+    # initialized in `on_locust_init`
+    funding_account = None
+
+    @classmethod
+    def get_next_id(cls):
+        cls.id_counter += 1
+        return cls.id_counter
+
+    @classmethod
+    def account_id(cls, id):
+        # Pseudo-random 6-digit prefix to spread the users in the state tree
+        prefix = str(hash(str(id)))[-6:]
+        return f"{prefix}_user{id}.{cls.funding_account.key.account_id}"
+
+    def __init__(self, environment):
+        super().__init__(environment)
+        host, port = self.host.split(":")
+        self.node = cluster.RpcNode(host, port)
+        self.id = NearUser.get_next_id()
+        self.account_id = NearUser.account_id(self.id)
+
+    def on_start(self):
+        """
+        Called once per user, creating the account on chain
+        """
+        self.account = Account(key.Key.from_random(self.account_id))
+        self.send_tx_retry(
+            CreateSubAccount(NearUser.funding_account,
+                             self.account.key,
+                             balance=5000.0))
+        self.account.refresh_nonce(self.node)
+
+    def send_tx(self, tx: Transaction):
+        """
+        Send a transaction and return the result, no retry attempted.
+        """
+        block_hash = base58.b58decode(self.node.get_latest_block().hash)
+        signed_tx = tx.sign_and_serialize(block_hash)
+
+        # doesn't work because it raises on status etc
+        # rpc_result = self.node.send_tx_and_wait(signed_tx, timeout=DEFAULT_TRANSACTION_TTL_SECONDS)
+
+        params = [base64.b64encode(signed_tx).decode('utf8')]
+        j = {
+            "method": "broadcast_tx_commit",
```

Comment:
Is waiting for `broadcast_tx_commit` to finish (rather than `broadcast_tx_async`) going to limit the test throughput much? Maybe if many NearUsers are spawned to send transactions, it's not a problem I guess. But if it is, maybe could be worth just sending off the tx here and checking the execution results elsewhere

Reply:
This is an important point that I want to tackle later. (Or maybe you want to look into it?)

But yeah, there is definitely a trade-off. Although I don't think test throughout is the main concern.

Locust is designed to make synchronous calls in each user and switch between them efficiently. Apparently it uses gevent under the hood which I believe makes HTTP calls behave in a asychrnous-IO way and switches efficiently between waiting sockets. (Case in point:  a single python thread is able to serve close to 1000 concurrent users without noticeable throttling, plus scaling to more python threads is also easy)


`broadcast_tx_commit`
pro:
- it is a synchronous API and works really well with the locust one greenlet pet user model

con:
- potential bottleneck on the RPC node that also needs to wait for the request to complete (doesn't look like I've hit that so far)
- we expect mainnet users to use `broadcast_tx_async`, so that would be more representative
- there is one open TCP connections for each user, which itself can become a bottleneck

So really, I just chose this because it's much simpler and it seem easily "good enough" for now.